[Captcha] Manually initiate a challenge for the specified user(s).

AAA3A-AAA3A opened the pull request:
### Type

- [ ] Bugfix
- [ ] Enhancement
- [X] New feature
- [ ] Documentation

### Description of the changes

Hello,
Here is the code to add the functionality to manually launch the captcha challenge for the specified user(s).
Simply issue the command "[p]captcha @user @user1 etc.".
Thanks in advance and have a nice day.

Kreusada commented:
Thanks for looking to contribute. As you can see, I've added the November milestone to your PR. This means I'll try and get onto it in the month of November, although I must take my exams first.

Kreusada commented:
Please run pre-commit on the necessary files.

Kreusada commented on captcha/events.py:
````diff
@@ -60,3 +60,25 @@ async def on_member_join(self, member: Member):
+    @commands.command()
+    async def captcha(self, ctx, *members: Member):
+        """Start a captcha challenge for the specified user"""
+        await ctx.send(
+            "The challenge has been launched for all members of the server specified in the command.\n"
+            + "```"
+            + ctx.message.content
+            + "```"
+        )
````
We can eradicate these lines entirely. Instead, we can just add a notice message and add typing until the final message has been sent.

```suggestion
        await ctx.send("Running Captcha challenges... this may take a while!")
        await ctx.trigger_typing()
```

Kreusada commented on captcha/events.py:
````diff
@@ -60,3 +60,25 @@ async def on_member_join(self, member: Member):
+        for i in members:
+            await self.runner(i)
+        await self.data.guild(ctx.guild).timeout.set(time)
+        await ctx.send(
+            "The challenge is over for all the users specified in the command. This does not mean that all users were not bots and therefore are still there.\n"
+            + "```"
+            + ctx.message.content
+            + "```"
+        )
````
I am not very happy with how this message has been constructed. Adding together strings is slower, and unconventional for the most part. You should use F-strings instead. Also, adding message content on the end is unnecessary. See my suggestion below:

```suggestion
        message = (
            "**The challenge has finished for the following members**\n"
            + ", ".join(member.name for member in members if not member.bot)
        )
        if any(member.bot for member in members):
            message += (
                "\n\n**The following members were not challenged because they were bots**\n"
                + ", ".join(member.name for member in members if member.bot)
            )
        await ctx.send(message)
```

Kreusada commented on captcha/events.py:
```diff
@@ -60,3 +60,25 @@ async def on_member_join(self, member: Member):
     async def on_member_remove(self, member: Member):
         await self.cleaner(member)
+
+    @commands.is_owner()
```
I think you're wanting `@commands.guildowner()` here.

```suggestion
    @commands.guildowner()
```

AAA3A-AAA3A replied:
Oh yes. I had thought for the bot owner only, but the server owner should be able to do it.

Kreusada reviewed (changes requested):
A few requested changes you should amend before I can consider merging this PR.

AAA3A-AAA3A commented:
I have made all the changes you suggested.
Should I use black on the files?
Thanks you in advance.

Kreusada commented:
Yes, please run black again on those changes

AAA3A-AAA3A commented:
I just tested everything. Everything works and no errors on the console.
Before using black on it, I test against a bot and a user at the same time.

AAA3A-AAA3A commented:
I have another idea which I may do when I have finished my own cog in progress. It would be to allow another setting that would allow a server owner to challenge someone who "looks" like a bot without necessarily testing everyone at their entrance.

AAA3A-AAA3A commented:
I reformatted the only file that had been modified using black and isort. It's fine.

AAA3A-AAA3A commented:
I have changed my PR a lot.
For the "[p]captcha" command, it will send an error message if the challenge already exists, rather than aborting everything.
Then I added a "[p]skipcaptcha" command which, as the name suggests, will stop a captcha challenge for the specified user(s).
I will make the PR change soon.
What do you think of it?

Kreusada commented on captcha/events.py:
```diff
@@ -18,7 +20,12 @@ class Listeners(MixinMeta, metaclass=ABCMeta):
         allowed = await self.basic_check(member)
         if allowed:
-            challenge = await self.create_challenge_for(member)
+            try:
+                challenge = await self.create_challenge_for(member)
+            except AlreadyHaveCaptchaError:
+                raise AlreadyHaveCaptchaError(
+                    "The user already have a captcha object running."
+                )
```
```suggestion
            challenge = await self.create_challenge_for(member)
```
This seems pointless to me, re-raising the same error. We can just allow it to raise like beforehand

Kreusada commented on captcha/events.py:
```diff
@@ -53,10 +60,129 @@ async def cleaner(self, member: Member):
+            await self.congratulation(challenge, roles)
+            await self.remove_temprole(challenge)
+
+            def ok_check(msg: str):
+                return f"✅ {msg}"
+
+            await self.send_or_update_log_message(
+                challenge.guild,
+                ok_check(f"Captcha skiped by {author.name} ({author.id})."),
```
```suggestion
            await self.send_or_update_log_message(
                challenge.guild,
                f"✅ Captcha skiped by {author.name} ({author.id}).",
```
We can just add it into the string instead of creating a function.

Kreusada commented on captcha/events.py:
```diff
@@ -53,10 +60,129 @@ async def cleaner(self, member: Member):
+            )
+            await self.send_or_update_log_message(
+                challenge.guild,
+                bold("Roles added, Captcha skiped."),
```
```suggestion
                bold("Roles added, Captcha skipped."),
```

Kreusada commented on captcha/events.py:
```diff
@@ -53,10 +60,129 @@ async def cleaner(self, member: Member):
+            roles_name = [role.name for role in roles]
+            try:
+                await challenge.member.send(
+                    f"Please contact the administrator of {challenge.guild.name} for obtaining "
+                    "access of the server, I was unable to add you the roles on the server.\nYou "
+                    f"should have obtained the following roles: "
+                    f"{humanize_list(roles_name) if roles_name else 'None.'}"
```
```suggestion
                    f"Please contact the administrator of {challenge.guild.name} in order to obtain "
                    "access to the server, I was unable to give you the roles on the server."
```
Don't need to list roles here

AAA3A-AAA3A replied:
It was already in the original, but I'm taking it out.

Kreusada commented on captcha/events.py:
```diff
@@ -53,10 +60,129 @@ async def cleaner(self, member: Member):
+                await challenge.channel.send(
+                    challenge.member.mention
+                    + ": "
+                    + f"Please contact the administrator of {challenge.guild.name} for obtaining "
+                    "access of the server, I was unable to add you the roles on the server.\nYou "
+                    f"should have obtained the following roles: "
+                    f"{humanize_list(roles_name) if roles_name else 'None.'}",
```
```suggestion
                    + f"Please contact the administrator of {challenge.guild.name} in order to obtain "
                    "access to the server, I was unable to give you the roles on the server."
```

Kreusada commented on captcha/events.py:
```diff
@@ -53,10 +60,129 @@ async def cleaner(self, member: Member):
+    @commands.guildowner()
+    @commands.command()
+    async def captcha(self, ctx, *members: Member):
+        """Start a captcha challenge for the specified user"""
```
```suggestion
        """Start a captcha challenge for the specified members."""
```

Kreusada commented on captcha/events.py:
```diff
@@ -53,10 +60,129 @@ async def cleaner(self, member: Member):
+    @commands.guildowner()
+    @commands.command(aliases=["bypasscaptcha"])
+    async def skipcaptcha(self, ctx, *members: Member):
+        """Cancel a captcha challenge for the specified user"""
```
```suggestion
        """Cancel a captcha challenge for the specified members."""
```

Kreusada reviewed (changes requested):
A lot needs to change in this PR, please take a look at the review. Also, please test your changes and make sure they work - I'm not doing that for you

AAA3A-AAA3A commented:
Hello,
I made all the requested changes, tested everything and ran isort and black on it.
Everything is good. Any other modification ?
Thanks in advance.

Kreusada commented:
please run black on your recent changes

Kreusada commented:
There's also undefined variables everywhere, I do not believe you have tested this. Please go away and test it properly

AAA3A-AAA3A commented:
I had already spent black and isort on it. You can see that I have uploaded again.
I really tested the commands and they work. I didn't go any further.
I'll look at the undefined variables.

AAA3A-AAA3A commented:
All undefined variables are corrected. That's all right. I ran black and isort on it.

Kreusada commented:
Unfortunately I have come to the decision that I would not like to see this feature merged into master. With dpy2 on the horizon, I'd much rather focus on fundamental implementation than feature ideas. You may maintain your own fork so that you can continue using these new commands for your instance of Red. Thank you very much for your willingness to contribute, and for your time dedicated to this PR.
